### PR TITLE
fix(src/index.js): add timeout to fastly request

### LINKED ISSUE #326: requests to fastly block and timeout after 60s
**Description**
image

### PR DESCRIPTION
**Pull request by MarquiseRosier**
currently request to fastly public-ip-list times out very frequently
this leads to action failure as the container has a 6000ms timeout
on generating an action response. This code adds timeout at 3000ms with action
failure; after retries at 500ms timeout intervals.

fix #326

Please ensure your pull request adheres to the following guidelines:
- [x] make sure to link the related issues in this description
- [x] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues
fix #326 

Thanks for contributing!

### CODE REVIEW

**Review by stefan-guggisberg — COMMENTED** [review]
> this leads to action failure as the container has a 6000ms timeout

AFAIU connections to blocking invocations are held open for 60s max. If a blocking action doesn't respond within 60s the invocation fails with an error.

File: test/utils.js [hunk 1]
```diff
@@ -9,10 +9,28 @@
...
+const { fetch, timeoutSignal, AbortError } = require('@adobe/helix-fetch');
 const assert = require('assert');
 
 function assertContains(actual, patterns) {
   patterns.map((expected) => assert.ok(new RegExp(expected).test(actual), `${actual} does not match ${expected}`));
 }
 
-module.exports = { assertContains };
+async function fetchAndRetry(url, duration = 0, defaultTimeout = 300) {
+  if (duration >= 1000) {
```
Comment (stefan-guggisberg) [inline test/utils.js:20]: In the description of this PR you wrote:

> This code adds timeout at 3000ms with action failure; after retries at 500ms timeout intervals.

That's not what the code is doing. The timeout/retry interval is 300ms and the fetch fails after 1000ms.

**Review by stefan-guggisberg — COMMENTED** [review]
Adding a timeout to the fetch call is certainly a good idea. 

However I don't think that retry logic (apart from the issues pointed out in my other comments) solves the particular issue at hand (#326). We've seen requests to `https://api.fastly.com/public-ip-list` failing with `ETIMEDOUT` errors after 60s repeatedly in the past (60s is the default timeout in the underlying socket implementation). The `https://api.fastly.com/public-ip-list` endpoint seems to have issues occasionally.  Retrying every 300ms wouldn't help in such a case.

I suggest we just add a timeout (e.g. 5s) to the fetch and fail if the fetch times out. No retry logic.

Furthermore we should find a solution where we don't need to rely on `https://api.fastly.com/public-ip-list`.

File: test/utils.js [hunk 2]
```diff
@@ -9,10 +9,28 @@
...
+    throw new Error(`Request to ${url} failing after multiple connection retries`);
+  }
+  let resp;
+  try {
+    resp = await fetch(url, { signal: timeoutSignal(defaultTimeout) });
+  } catch (err) {
+    if (err instanceof AbortError) {
+      fetchAndRetry(url, duration + defaultTimeout);
```
[the remark this answers is not in the document]
Comment (MarquiseRosier) [reply test/utils.js:28]: It's a default parameter there! the timeout doesn't need to be passed again!

**Comment by trieloff** [comment]
> Furthermore we should find a solution where we don't need to rely on `https://api.fastly.com/public-ip-list`.

1. as part of the CI build, download the list, store it as a JSON file, and commit it back (using `[ci skip]`)
2. `require` the JSON file in our code

we will need to re-build when the list changes, but this is infrequent.

File: test/utils.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefan-guggisberg) [reply test/utils.js:28]: > It's a default parameter there! 

That's exactly the point :-) If you initially call `fetchAndRetry` with a timeout of `1000` the timeout won't be passed to the next recursion level.

File: .circleci/config.yml [hunk 3]
```diff
@@ -28,6 +28,12 @@ commands:
     - run:
         name: prepare test git user
         command: git config --global user.email "circleci@example.com" && git config --global user.name "CircleCi Build"
+    - run:
+        name: download fastly public ip list
+        command: curl "https://api.fastly.com/public-ip-list" > ip-list.json
+    - run: 
+        name: commit fastly public ip list
+        command: git add ip-list.json; git commit -m "[ci skip]";
```
Comment (tripodsan) [inline .circleci/config.yml:36]:
```suggestion
        command: git add ip-list.json; git commit -m "chore: updating ip-list [skip ci]";
```

File: src/embed.js [hunk 4]
```diff
@@ -24,6 +26,26 @@ matchers.push(unsplash);
...
 matchers.push(spotify);
 
+/**
+ * function reads file and returns json object
+ * @param {string} file name of file to be loaded
+ * @param {object} log handle on logging object
+ * @returns json object of fastly public ips
+ */
+async function getIPList(file, log = console) {
+  const fd = path.resolve(file);
```
Comment (tripodsan) [inline src/embed.js:36]: why `fd` ?

**Review by tripodsan — APPROVED** [review]

**Comment by stefan-guggisberg** [comment]
the build is still failing.

**Comment by tripodsan** [comment]
> the build is still failing.

argh.... :-)

**Review by stefan-guggisberg — APPROVED** [review]

File: src/embed.js (hunk 4 above)
Comment (trieloff) [reply src/embed.js:36]: And wouldn't a plain `require` work, too?

**Review by trieloff — APPROVED** [review]

**Pull request merged**